### PR TITLE
Zeppelin 1307 - Implement notebook revision in Zeppelinhub repo

### PR DESCRIPTION
**Pull request by anthonycorbacho**
### What is this PR for?
Implement versioning in ZeppelinHub notebook storage.

### What type of PR is it?
Improvement

### Todos
* [x] - Implement Versioning API

### What is the Jira issue?
 * [ZEPPELIN-1307](https://issues.apache.org/jira/browse/ZEPPELIN-1307)

### How should this be tested?
Edit `zeppelin-env.sh` and add `org.apache.zeppelin.notebook.repo.zeppelinhub.ZeppelinHubRepo` in `ZEPPELIN_NOTEBOOK_STORAGE`.

### Questions:
* Does the licenses files need update? NO
* Is there breaking changes for older versions? NO
* Does this needs documentation? NO

### CODE REVIEW

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/zeppelinhub/rest/ZeppelinhubRestApiHandler.java [hunk 1]
```diff
@@ -128,17 +128,17 @@ public String asyncGet(String argument) throws IOException {
       response = listener.get(30, TimeUnit.SECONDS);
     } catch (InterruptedException | TimeoutException | ExecutionException e) {
       LOG.error("Cannot perform Get request to ZeppelinHub", e);
-      throw new IOException("Cannot load note from ZeppelinHub", e);
+      throw new IOException("Cannot perform Get request to ZeppelinHub", e);
     }
 
     int code = response.getStatus();
     if (code == 200) {
-      try (InputStream responseContent = listener.getInputStream()) {
```
Comment (khalidhuseynov) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/zeppelinhub/rest/ZeppelinhubRestApiHandler.java]: here [try-with-resource statements](https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html) (and [here](https://www.mkyong.com/java/try-with-resources-example-in-jdk-7/)) is automatically closable, and it's syntactic sugar for what you did, so basically they're equivalent and no need to change

Comment (anthonycorbacho) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/zeppelinhub/rest/ZeppelinhubRestApiHandler.java]: Sorry i didnt know that InputStream implement Closable, i will revert it.

**Comment by anthonycorbacho** [comment]
ready for review

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/zeppelinhub/ZeppelinHubRepo.java [hunk 2]
```diff
@@ -191,20 +193,45 @@ public void close() {
   @Override
   public Revision checkpoint(String noteId, String checkpointMsg, AuthenticationInfo subject)
       throws IOException {
-    // Auto-generated method stub
-    return null;
+    if (StringUtils.isBlank(noteId)) {
+      return null;
```
Comment (bzz) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/zeppelinhub/ZeppelinHubRepo.java]: To be consistent, may be we should create `Revision.EMPTY` or something like that?

Comment (anthonycorbacho) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/zeppelinhub/ZeppelinHubRepo.java]: This might be a subject of another pr because i am following the convention
of the other notebookrepo.

This is out of the scope of this pr.

**Comment by anthonycorbacho** [comment]
**L**ets **G**et **T**o **M**erge

**Comment by anthonycorbacho** [comment]
@bzz Regarding your comment, I couldnt agree more, i am not a big fan of returning `null`. I created an issue in [JIRA - 1352](https://issues.apache.org/jira/browse/ZEPPELIN-1352) and I will address it ASAP after merging this PR

**Comment by anthonycorbacho** [comment]
merging if no more discussion

**Comment by khalidhuseynov** [comment]
somehow i keep getting this error when doing checkpoint: 
```
ERROR [2016-08-22 14:33:04,618] ({qtp787867107-76} NotebookServer.java[onMessage]:257) - Can't handle message
java.lang.NullPointerException
	at org.eclipse.jetty.client.HttpClient.send(HttpClient.java:521)
	at org.eclipse.jetty.client.HttpRequest.send(HttpRequest.java:694)
	at org.eclipse.jetty.client.HttpRequest.send(HttpRequest.java:678)
	at org.apache.zeppelin.notebook.repo.zeppelinhub.rest.ZeppelinhubRestApiHandler.sendToZeppelinHub(ZeppelinhubRestApiHandler.java:157)
	at org.apache.zeppelin.notebook.repo.zeppelinhub.rest.ZeppelinhubRestApiHandler.asyncPutWithResponseBody(ZeppelinhubRestApiHandler.java:125)
	at org.apache.zeppelin.notebook.repo.zeppelinhub.ZeppelinHubRepo.checkpoint(ZeppelinHubRepo.java:201)
	at org.apache.zeppelin.notebook.repo.NotebookRepoSync.checkpoint(NotebookRepoSync.java:334)
	at org.apache.zeppelin.notebook.Notebook.checkpointNote(Notebook.java:358)
	at org.apache.zeppelin.socket.NotebookServer.checkpointNotebook(NotebookServer.java:1182)
	at org.apache.zeppelin.socket.NotebookServer.onMessage(NotebookServer.java:233)
	at org.apache.zeppelin.socket.NotebookSocket.onWebSocketText(NotebookSocket.java:56)
	at org.eclipse.jetty.websocket.common.events.JettyListenerEventDriver.onTextMessage(JettyListenerEventDriver.java:128)
	at org.eclipse.jetty.websocket.common.message.SimpleTextMessage.messageComplete(SimpleTextMessage.java:69)
	at org.eclipse.jetty.websocket.common.events.AbstractEventDriver.appendMessage(AbstractEventDriver.java:65)
	at org.eclipse.jetty.websocket.common.events.JettyListenerEventDriver.onTextFrame(JettyListenerEventDriver.java:122)
	at org.eclipse.jetty.websocket.common.events.AbstractEventDriver.incomingFrame(AbstractEventDriver.java:161)
	at org.eclipse.jetty.websocket.common.WebSocketSession.incomingFrame(WebSocketSession.java:309)
	at org.eclipse.jetty.websocket.common.extensions.ExtensionStack.incomingFrame(ExtensionStack.java:214)
	at org.eclipse.jetty.websocket.common.Parser.notifyFrame(Parser.java:220)
	at org.eclipse.jetty.websocket.common.Parser.parse(Parser.java:258)
	at org.eclipse.jetty.websocket.common.io.AbstractWebSocketConnection.readParse(AbstractWebSocketConnection.java:632)
	at org.eclipse.jetty.websocket.common.io.AbstractWebSocketConnection.onFillable(AbstractWebSocketConnection.java:480)
	at org.eclipse.jetty.io.AbstractConnection$2.run(AbstractConnection.java:544)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:635)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:555)
	at java.lang.Thread.run(Thread.java:745)
```
not sure what's the problem though, let me know if doing something wrong

**Comment by anthonycorbacho** [comment]
@khalidhuseynov fixed the issue, thank for testing

**Comment by khalidhuseynov** [comment]
tested and works well now! LGTM
